docs/sdk-spec.markdown

```diff
@@ -0,0 +1,256 @@
+this can be the device path, like `/dev/disk2`, or `\\\\.\\PHYSICALDRIVE3`.
+
+### Scanner
+
+A scanner is an entity that inspects the real world hoping to find a set of
+*destinations*. Scanners are singleton objects that look like this:
+
+#### `EventEmitter .start(Object options)`
+
+This function starts the scanner.
+
+- `Number options.times` (defaults to `Infinity`): The number of scans to run.
+  When the scanner reaches this number, the scanner stops
+- `Number options.delay (defaults to `0`)`: The amount of milliseconds to wait
+  between scans
+- `Object options.extra`: Scanner-specific extra options
+
+The returned `EventEmitter` should emit these events:
+
+- `start`: A scan phase has just started. The handler argument is the number of
+  times the scanner will start
+- `stop`: The scanner will stop scanning
+- `done`: The scan phase ended. The handler argument is an array of
+  *destinations*
+- `found`: Emitted potentially various times during a single scan phase,
+  passing a single *destination* at a time. This event is useful for scanners
+  that can go incrementally
+- `wait`: Emitted when the scanner enters a waiting phase. The handler argument
+  is the number of milliseconds the waiting will last
+- `error`: An error occurred. In such case, the scanner should continue with
+  the next scan loop, unless the user explicitly stops it on the `error`
+  handler
+
+The scanner implementation may emit custom events, as long as they include an
+'extra:' prefix.
+
+#### `EventEmitter .stop()`
+
+This function stops a scanner.
+
+The returned `EventEmitter` should emit these events:
+
+- `done`: The stop action finished successfully
+- `error`: An error occurred, and the scanner might be still running
+
+The scanner implementation may emit custom events, as long as they include an
+`extra:` prefix.
+
+#### `Boolean .isScanning()`
+
+This function checks if a scanner is active. This function should return `true`
+even if the scanner is in a waiting phase.
+
+#### `EventEmitter .getAll(Object options)`
+
+This function retrieves the list of destinations found by the scanner.
+
+- If the scanner is active, then it will return the last found list of
+  destinations
+- If the scanner is inactive, then it will return an empty array
+
+The returned `EventEmitter` should emit these events:
+
+- `done`: The retrieval was successful, and the handler argument is an array of
+  destinations
+- `error`: An error occurred
+
+The supported options are scanner specific. In the context of local drives, one
+option could be a `safe` boolean that would cause `.getAll()` to filter out
+fixed and system drives.
+
+#### `EventEmitter .get(String id)`
+
+This function retrieves a single destination by its id.
+
+- If the scanner is active, and the id exists, then it will return the
+  destination
+- If the scanner is active, and the id exists, then it will return `null`
+- If the scanner is not running, then it will return `null`
```

Comment:
Yeah, in the context of Etcher, all scanners will be continuously running, and the client, which is the current dialog to select the destination, will query it as needed.

We could add a way to allow the scanner to run once and stop if that's needed.